File: packages/merkle-tree/src/test/metadata.test.ts
import fs from 'fs/promises';
import * as crypto from 'crypto';
import { 
    IMerkleTree,
    HashedItem,
    addItem,
    updateItem,
    deleteItem,
    saveTree,
    loadTree,
    createTree,
    buildMerkleTree
} from '../lib/merkle-tree';
import { FileStorage } from 'storage';

describe('Merkle Tree Metadata', () => {
    const TEST_FILE_PATH = './test-tree-metadata.bin';
    
    beforeEach(() => {
        jest.useFakeTimers();
    });
    
    /**
     * Helper function to create a file hash with a given name and content
     */
    function createHashedItem(name: string, content: string = name): HashedItem {
        // Create a proper 32-byte SHA-256 hash
        const hash = crypto.createHash('sha256').update(content).digest();
        return {
            name,
            hash,
            length: content.length,
            lastModified: new Date(),
        };
    }

    /**
     * Helper function to build a tree with the given file names
     */
    function buildTree(fileNames: string[]): IMerkleTree<any>{
        let merkleTree = createTree("12345678-1234-5678-9abc-123456789abc");
        
        for (const fileName of fileNames) {
            const fileHash = createHashedItem(fileName);
            merkleTree = addItem(merkleTree, fileHash);
        }

        if (!merkleTree) {
            throw new Error('Failed to build the tree');
        }

        merkleTree.dirty = false;
        merkleTree.merkle = buildMerkleTree(merkleTree.sort);
        
        return merkleTree;
    }
    
    // Clean up test file after each test
    afterEach(async () => {
        try {
            await fs.access(TEST_FILE_PATH);
            await fs.unlink(TEST_FILE_PATH);
        } catch (error) {
            // File doesn't exist, nothing to do
        }
    });
    
    test('should initialize metadata when creating a new tree', () => {
        const originalTree = buildTree(['A', 'B']);
        
        // Check that the metadata exists
        // Verify UUID is properly formatted
        expect(originalTree.id).toMatch(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/);
        
        // Verify node and file counts
        expect(originalTree.sort?.nodeCount).toBe(originalTree.sort?.nodeCount || 0);
        expect(originalTree.sort?.leafCount).toBeGreaterThan(0);
    });
    
    test('should update metadata when modifying the tree', () => {
        // Create initial tree
        let tree = buildTree(['A', 'B']);
        const originalMetadata = { id: tree.id, nodeCount: tree.sort?.nodeCount, leafCount: tree.sort?.leafCount };
        
        // Let time pass
        jest.advanceTimersByTime(1000);
        
        // Add a new file
        const fileHashC = createHashedItem('C');
        tree = addItem(tree, fileHashC);
        
        // Check that metadata was updated
        // UUID should remain the same
        expect(tree.id).toEqual(originalMetadata.id);
        
        // Counts should be updated
        expect(tree.sort?.nodeCount).toBe(tree.sort?.nodeCount || 0);
        expect(tree.sort?.leafCount).toBeGreaterThan(0);
    });
    
    test('should update metadata when updating a file', () => {
        // Create initial tree
        let tree = buildTree(['A', 'B']);
        const originalMetadata = { id: tree.id, nodeCount: tree.sort?.nodeCount, leafCount: tree.sort?.leafCount };
        
        // Let time pass
        jest.advanceTimersByTime(1000);
        
        // Update file A
        const updatedHashedItemA = createHashedItem('A', 'modified content');
        updateItem(tree, updatedHashedItemA);
        
        // Check that metadata was updated
        // UUID should remain the same
        expect(tree.id).toEqual(originalMetadata.id);
        
        // Counts should remain the same (updating doesn't add nodes)
        expect(tree.sort?.nodeCount).toBe(originalMetadata.nodeCount);
        expect(tree.sort?.leafCount).toBe(originalMetadata.leafCount);
    });
    
    test('should update metadata when deleting a file', () => {
        // Create initial tree
        let tree = buildTree(['A', 'B', 'C']);
        const originalMetadata = { id: tree.id, nodeCount: tree.sort?.nodeCount, leafCount: tree.sort?.leafCount };
        
        // Let time pass
        jest.advanceTimersByTime(1000);
        
        // Delete file B
        deleteItem(tree, 'B');
        
        // Check that metadata was updated
        // UUID should remain the same
        expect(tree.id).toEqual(originalMetadata.id);
        
        // Counts should decrease (hard delete removes nodes)
        expect(tree.sort?.nodeCount || 0).toBeLessThan(originalMetadata.nodeCount || 0);
        expect(tree.sort?.leafCount).toBe((originalMetadata.leafCount || 0) - 1);
    });
    
    test('should save and load metadata with V2 format', async () => {
        // Create a tree with metadata
        const originalTree = buildTree(['A', 'B', 'C']);
        // Save the tree to a file
        await saveTree(TEST_FILE_PATH, originalTree, new FileStorage(""));
        
        // Load the tree from the file
        const loadedTree = (await loadTree(TEST_FILE_PATH, new FileStorage("")))!;
        
        // All metadata fields should match
        expect(loadedTree.id).toEqual(originalTree.id);
        expect(loadedTree.sort?.nodeCount).toEqual(originalTree.sort?.nodeCount);
        expect(loadedTree.sort?.leafCount).toEqual(originalTree.sort?.leafCount);
    });
    
    // Tests removed - createDefaultMetadata and updateMetadata no longer exist
    // Metadata (totalNodes, totalFiles, totalSize) now comes directly from the sort tree root
});

File: packages-zig/merkle-tree-zig/src/test/metadata.test.zig
//
// Tree metadata tests (port of src/test/metadata.test.ts).
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const memory_storage = @import("memory-storage.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;

//
// Helper function to build a tree with the given file names
//
fn buildTree(allocator: std.mem.Allocator, fileNames: []const []const u8) !IMerkleTree {
    var merkleTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    for (fileNames) |fileName| {
        merkleTree = try merkle_tree.addItem(allocator, &merkleTree, try merkle_verify.createSha256HashedItem(allocator, fileName, fileName, fileName.len));
    }
    merkleTree.dirty = false;
    merkleTree.merkle = try merkle_tree.buildMerkleTree(allocator, merkleTree.sort);
    return merkleTree;
}

//
// Returns true when text has the lowercase UUID format.
//
fn isUuidFormat(text: []const u8) bool {
    if (text.len != 36) {
        return false;
    }
    for (text, 0..) |character, index| {
        if (index == 8 or index == 13 or index == 18 or index == 23) {
            if (character != '-') {
                return false;
            }
        }
        else if (!std.ascii.isDigit(character) and !(character >= 'a' and character <= 'f')) {
            return false;
        }
    }
    return true;
}

test "should initialize metadata when creating a new tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const originalTree = try buildTree(arena.allocator(), &.{ "A", "B" });
    try std.testing.expect(isUuidFormat(originalTree.id));
    try std.testing.expect(originalTree.sort.?.leafCount > 0);
}

test "should update metadata when modifying the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTree(allocator, &.{ "A", "B" });
    const originalId = tree.id;
    tree = try merkle_tree.addItem(allocator, &tree, try merkle_verify.createSha256HashedItem(allocator, "C", "C", 1));
    try std.testing.expectEqualStrings(originalId, tree.id);
    try std.testing.expect(tree.sort.?.leafCount > 0);
}

test "should update metadata when updating a file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTree(allocator, &.{ "A", "B" });
    const originalId = tree.id;
    const originalNodeCount = tree.sort.?.nodeCount;
    const originalLeafCount = tree.sort.?.leafCount;
    _ = try merkle_tree.updateItem(&tree, try merkle_verify.createSha256HashedItem(allocator, "A", "modified content", 16));
    try std.testing.expectEqualStrings(originalId, tree.id);
    try std.testing.expectEqual(originalNodeCount, tree.sort.?.nodeCount);
    try std.testing.expectEqual(originalLeafCount, tree.sort.?.leafCount);
}

test "should update metadata when deleting a file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTree(allocator, &.{ "A", "B", "C" });
    const originalId = tree.id;
    const originalNodeCount = tree.sort.?.nodeCount;
    const originalLeafCount = tree.sort.?.leafCount;
    try merkle_tree.deleteItem(allocator, &tree, "B");
    try std.testing.expectEqualStrings(originalId, tree.id);
    try std.testing.expect(tree.sort.?.nodeCount < originalNodeCount);
    try std.testing.expectEqual(originalLeafCount - 1, tree.sort.?.leafCount);
}

test "should save and load metadata with V2 format" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const originalTree = try buildTree(allocator, &.{ "A", "B", "C" });
    var storage = memory_storage.MemoryStorage.init(allocator);
    try merkle_tree.saveTree(allocator, std.testing.io, "test-tree-metadata.bin", &originalTree, storage.asStorage(), "FTRE");
    const loadedTree = (try merkle_tree.loadTree(allocator, std.testing.io, "test-tree-metadata.bin", storage.asStorage(), "FTRE")).?;
    try std.testing.expectEqualStrings(originalTree.id, loadedTree.id);
    try std.testing.expectEqual(originalTree.sort.?.nodeCount, loadedTree.sort.?.nodeCount);
    try std.testing.expectEqual(originalTree.sort.?.leafCount, loadedTree.sort.?.leafCount);
}
